File: sdks/python/tests/test_exporter.py
"""Tests for etrace._exporter — SpanExporter protocol and built-in implementations."""

from __future__ import annotations

from etrace._exporter import (
    ConsoleExporter,
    InMemoryExporter,
    SpanExportResult,
)
from etrace._types import Span, TraceKind, TraceStatus


def _make_span(name: str = "test", kind: TraceKind = TraceKind.TOOL) -> Span:
    return Span(
        trace_id="0" * 32,
        span_id="0" * 16,
        name=name,
        kind=kind,
        status=TraceStatus.OK,
        started_at="2025-01-01T00:00:00Z",
        ended_at="2025-01-01T00:00:01Z",
        duration_ns=1_000_000_000,
    )


# ── SpanExportResult ──────────────────────────────────────────────────────────


class TestSpanExportResult:
    def test_success_failure_values(self):
        assert SpanExportResult.SUCCESS.value == "success"
        assert SpanExportResult.FAILURE.value == "failure"


# ── InMemoryExporter ──────────────────────────────────────────────────────────


class TestInMemoryExporter:
    def test_export_accumulates_spans(self):
        exporter = InMemoryExporter()
        s1, s2 = _make_span("a"), _make_span("b")
        assert exporter.export([s1, s2]) == SpanExportResult.SUCCESS
        assert exporter.get_finished_spans() == [s1, s2]

    def test_export_returns_success(self):
        exporter = InMemoryExporter()
        assert exporter.export([_make_span()]) == SpanExportResult.SUCCESS

    def test_export_empty_batch(self):
        exporter = InMemoryExporter()
        assert exporter.export([]) == SpanExportResult.SUCCESS
        assert exporter.get_finished_spans() == []

    def test_get_finished_spans_returns_copy(self):
        exporter = InMemoryExporter()
        s = _make_span()
        exporter.export([s])
        spans = exporter.get_finished_spans()
        spans.clear()
        assert exporter.get_finished_spans() == [s]

    def test_clear_removes_all_spans(self):
        exporter = InMemoryExporter()
        exporter.export([_make_span(), _make_span()])
        exporter.clear()
        assert exporter.get_finished_spans() == []

    def test_force_flush_returns_true(self):
        exporter = InMemoryExporter()
        assert exporter.force_flush() is True
        assert exporter.force_flush(timeout_millis=1000) is True

    def test_shutdown_is_safe(self):
        exporter = InMemoryExporter()
        exporter.export([_make_span()])
        exporter.shutdown()
        assert exporter.get_finished_spans() == [s for s in exporter.get_finished_spans()]

    def test_multiple_exports_accumulate(self):
        exporter = InMemoryExporter()
        exporter.export([_make_span("1")])
        exporter.export([_make_span("2")])
        exporter.export([_make_span("3")])
        assert len(exporter.get_finished_spans()) == 3

    def test_spans_preserve_all_fields(self):
        exporter = InMemoryExporter()
        s = _make_span()
        s.model = "gpt-4o"
        s.provider = "openai"
        s.input = "hello"
        s.output = "world"
        exporter.export([s])
        exported = exporter.get_finished_spans()[0]
        assert exported.model == "gpt-4o"
        assert exported.provider == "openai"
        assert exported.input == "hello"
        assert exported.output == "world"


# ── ConsoleExporter ───────────────────────────────────────────────────────────


class TestConsoleExporter:
    def test_export_writes_to_stdout(self, capsys):
        exporter = ConsoleExporter()
        s = _make_span("my_span")
        exporter.export([s])
        captured = capsys.readouterr()
        assert "my_span" in captured.out

    def test_export_returns_success(self, capsys):
        exporter = ConsoleExporter()
        assert exporter.export([_make_span()]) == SpanExportResult.SUCCESS

    def test_export_multiple_spans(self, capsys):
        exporter = ConsoleExporter()
        exporter.export([_make_span("span_a"), _make_span("span_b")])
        captured = capsys.readouterr()
        assert "span_a" in captured.out
        assert "span_b" in captured.out

    def test_custom_output_stream(self):
        import io

        buf = io.StringIO()
        exporter = ConsoleExporter(out=buf)
        exporter.export([_make_span("custom_out")])
        assert "custom_out" in buf.getvalue()

    def test_force_flush_returns_true(self):
        exporter = ConsoleExporter()
        assert exporter.force_flush() is True

    def test_shutdown_is_safe(self):
        exporter = ConsoleExporter()
        exporter.shutdown()


# ── Custom exporter (protocol compliance) ─────────────────────────────────────


class TestCustomExporterProtocol:
    def test_custom_exporter_works(self):
        """Any object with export/shutdown/force_flush satisfies the protocol."""

        class MyExporter:
            def __init__(self):
                self.exported = []

            def export(self, spans):
                self.exported.extend(spans)
                return SpanExportResult.SUCCESS

            def shutdown(self):
                pass

            def force_flush(self, timeout_millis=30000):
                return True

        exp = MyExporter()
        s = _make_span()
        exp.export([s])
        assert len(exp.exported) == 1

File: sdks/typescript/tests/test_exporter.ts
/**
 * Tests for SpanExporter protocol, InMemoryExporter, ConsoleExporter.
 *
 * Matches Python test_exporter.py (17 tests).
 */
import { describe, it, expect } from "vitest";
import { InMemoryExporter, ConsoleExporter, SpanExportResult } from "../src/exporter.js";
import type { Span } from "../src/types.js";

function makeSpan(overrides: Partial<Span> = {}): Span {
  return {
    traceId: crypto.randomUUID().replace(/-/g, ""),
    spanId: crypto.randomUUID().replace(/-/g, "").slice(0, 16),
    name: "test",
    kind: "custom",
    status: "ok",
    startedAt: new Date().toISOString(),
    ...overrides,
  };
}

describe("SpanExportResult", () => {
  it("has SUCCESS=0 and FAILED=1", () => {
    expect(SpanExportResult.SUCCESS).toBe(0);
    expect(SpanExportResult.FAILED).toBe(1);
  });
});

describe("InMemoryExporter", () => {
  it("exports spans and returns SUCCESS", () => {
    const exp = new InMemoryExporter();
    const span = makeSpan();
    const result = exp.export([span]);
    expect(result).toBe(SpanExportResult.SUCCESS);
  });

  it("stores exported spans", () => {
    const exp = new InMemoryExporter();
    const s1 = makeSpan({ name: "a" });
    const s2 = makeSpan({ name: "b" });
    exp.export([s1, s2]);
    expect(exp.getFinishedSpans()).toHaveLength(2);
    expect(exp.getFinishedSpans()[0].name).toBe("a");
    expect(exp.getFinishedSpans()[1].name).toBe("b");
  });

  it("returns a copy from getFinishedSpans", () => {
    const exp = new InMemoryExporter();
    exp.export([makeSpan()]);
    const copy = exp.getFinishedSpans();
    copy.push(makeSpan({ name: "extra" }));
    expect(exp.getFinishedSpans()).toHaveLength(1);
  });

  it("clear removes all spans", () => {
    const exp = new InMemoryExporter();
    exp.export([makeSpan()]);
    exp.clear();
    expect(exp.getFinishedSpans()).toHaveLength(0);
  });

  it("shutdown clears all spans", () => {
    const exp = new InMemoryExporter();
    exp.export([makeSpan()]);
    exp.shutdown();
    expect(exp.getFinishedSpans()).toHaveLength(0);
  });

  it("forceFlush returns true", () => {
    const exp = new InMemoryExporter();
    expect(exp.forceFlush()).toBe(true);
  });

  it("handles multiple sequential exports", () => {
    const exp = new InMemoryExporter();
    exp.export([makeSpan({ name: "a" })]);
    exp.export([makeSpan({ name: "b" })]);
    exp.export([makeSpan({ name: "c" })]);
    expect(exp.getFinishedSpans()).toHaveLength(3);
  });

  it("handles empty export", () => {
    const exp = new InMemoryExporter();
    const result = exp.export([]);
    expect(result).toBe(SpanExportResult.SUCCESS);
    expect(exp.getFinishedSpans()).toHaveLength(0);
  });
});

describe("ConsoleExporter", () => {
  it("exports spans and returns SUCCESS", () => {
    const exp = new ConsoleExporter();
    const result = exp.export([makeSpan()]);
    expect(result).toBe(SpanExportResult.SUCCESS);
  });

  it("shutdown does not throw", () => {
    const exp = new ConsoleExporter();
    expect(() => exp.shutdown()).not.toThrow();
  });
});
